File: src/main/java/com/jme/asset/store/service/AssetCategoryService.java
package com.jme.asset.store.service;

import com.jme.asset.store.db.entity.asset.AssetCategoryEntity;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * The service to work with asset category.
 *
 * @author Yunkevich Andrei
 */
public interface AssetCategoryService {

    /**
     * Create category
     *
     * @param name        the category name
     * @param description the category description
     * @param parent      the category parent
     */
    void addCategory(@NotNull String name, @Nullable String description, @Nullable AssetCategoryEntity parent);

    /**
     * Find asset category by id
     *
     * @param id the category id
     * @return the asset category
     */
    @Nullable AssetCategoryEntity load(long id);

    /**
     * Remove asset category
     *
     * @param category the category of asset
     */
    void removeCategory(@NotNull AssetCategoryEntity category);

    /**
     * Add child to children of asset category
     *
     * @param category      the asset category
     * @param childCategory the child of asset category
     */
    void addChild(@NotNull AssetCategoryEntity category, @NotNull AssetCategoryEntity childCategory);

    /**
     * Remove child from children of asset category
     *
     * @param category      the asset category
     * @param childCategory the child of asset category
     */
    void removeChild(@NotNull AssetCategoryEntity category, @NotNull AssetCategoryEntity childCategory);

    /**
     * Get asset categories
     *
     * @return the list of asset categories
     */
    @NotNull List<AssetCategoryEntity> getCategories();

    /**
     * Get children.
     *
     * @param parent the parent category.
     * @return the children.
     */
    @NotNull List<AssetCategoryEntity> getChildren(@NotNull AssetCategoryEntity parent);

    /**
     * Get category and all children by category id
     *
     * @param categoryId category id
     * @return list category
     */
    @NotNull List<AssetCategoryEntity> getCategoryAndAllChildren(@NotNull AssetCategoryEntity categoryId);

}

File: front-end/src/service/asset-category.service.ts
import {Injectable} from '@angular/core';
import {AssetCategoryEntity} from "../model/entity/asset-category-entity";
import {Http, RequestOptions, Headers} from "@angular/http";
import {Utils} from "../util/utils";
import {AssetCategoryParam} from "../model/category/asset-category-param";

import {SecurityService} from "./security.service";


/**
 * The asset category service.
 *
 * @author Denis Lesheniuk.
 */
@Injectable()
export class AssetCategoryService {

  /**
   * The url to categories controller.
   *
   * @type {string}
   */
  private static readonly API_ASSETS_CATEGORIES = "api/asset-categories";

  constructor(private readonly http: Http, private readonly securityService: SecurityService) {
  }

  /**
   * The function to get all categories from server.
   *
   * @returns {Promise<AssetCategoryEntity[]>} the all categories.
   */
  public getCategories(): Promise<AssetCategoryEntity[]> {
    var options = new RequestOptions({headers: new Headers()});
    this.securityService.appendAccessToken(options);
    return this.http.get(AssetCategoryService.API_ASSETS_CATEGORIES, options)
      .toPromise()
      .then(response => {
        let body = response.json();
        return body;
      })
      .catch(error => Utils.handleError(error));
  }

  /**
   * The function to put the category to the server.
   *
   * @param {AssetCategoryParam} params the asset category param.
   * @param {(message: string, result: boolean) => void} handler the handler result of putting.
   */
  public addCategory(params: AssetCategoryParam,
                     handler: (message: string, result: boolean) => void) {
    var options = new RequestOptions({headers: new Headers()});
    this.securityService.appendAccessToken(options);
    this.http.put(AssetCategoryService.API_ASSETS_CATEGORIES + "/put", params, options)
      .toPromise()
      .then(response => handler(null, true))
      .catch(error => Utils.handleErrorMessageJson(error, (ex: string) => handler(ex, false)));
  }

  /**
   * The function to remove the category from server.
   *
   * @param {string} id the category id.
   * @param {(message: string, result: boolean) => void} handler the handler result of removing.
   */
  public removeCategory(id: string, handler: (message: string, result: boolean) => void) {
    var options = new RequestOptions({headers: new Headers()});
    this.securityService.appendAccessToken(options);
    this.http.delete(AssetCategoryService.API_ASSETS_CATEGORIES + '/' + id, options)
      .toPromise()
      .then(response => handler(null, true))
      .catch(error => Utils.handleErrorMessageJson(error, (ex: string) => handler(ex, false)));
  }

  /**
   * Get children.
   *
   * @param {string} id the parent id.
   * @returns {Promise<AssetCategoryEntity[]>} result of getting children.
   */
  public getChildren(id: string): Promise<AssetCategoryEntity[]> {
    var options = new RequestOptions({headers: new Headers()});
    this.securityService.appendAccessToken(options);
    return this.http.get(AssetCategoryService.API_ASSETS_CATEGORIES + "/" + id, options)
      .toPromise()
      .then(response => {
        let body = response.json();
        return body;
      })
      .catch(error => Utils.handleError(error));
  }
}
